File: safety_check.py
import sys
import os
import pypulseq as pp
import matplotlib.pyplot as plt

# Predefined acoustic resonance frequencies (example values - adjust as needed)
DEFAULT_ACOUSTIC_RESONANCES_3T = [ # For our Siemens 3T Prisma
    {"frequency": 590, "bandwidth": 100},
    {"frequency": 1140, "bandwidth": 220}
]

DEFAULT_ACOUSTIC_RESONANCES_7T = [ # For our Siemens 3T Prisma
    {"frequency": 550, "bandwidth": 100},
    {"frequency": 1100, "bandwidth": 300}
]

def extract_md5_hash(seq_file):
    """
    Extract the MD5 hash from the last line of the .seq file
    """
    try:
        with open(seq_file, 'r') as f:
            lines = f.readlines()
            # Look for the hash in the last few lines
            for line in reversed(lines):
                line = line.strip()
                if line.startswith("Hash "):
                    # Extract the hash part after "Hash "
                    hash_value = line[5:]  # Remove "Hash " prefix
                    return hash_value
        return "Hash not found"
    except Exception as e:
        print(f"Error reading seq file for hash: {e}")
        return "Hash extraction failed"

def check_PNS(seq, gradFile, seq_file):
    """
    Plots the PNS level of the sequence using the gradient information in gradFile and throws an error if PNS level is exceeded
    """
    # Check if gradient file exists
    gradient_file = os.path.join(os.path.dirname(__file__), gradFile)
    if os.path.exists(gradient_file):
        print(f"Gradient file found: {gradFile}, using it for PNS check.")
    else:
        print(f"Warning: Gradient file '{gradFile}' not found. Skipping PNS check.")
        return

    # Extract MD5 hash for the title
    md5_hash = extract_md5_hash(seq_file)

    pns_ok, pns_n, pns_c, tpns = seq.calculate_pns(gradient_file, do_plots=True)

    # Add title with MD5 hash
    plt.suptitle(f"PNS Check - Hash: {md5_hash}", fontsize=14, fontweight='bold')

    # Save the plot as PNG
    output_filename = f"pns_check_{os.path.basename(seq_file).replace('.seq', '')}.png"
    plt.savefig(output_filename, dpi=300, bbox_inches='tight')
    print(f"PNS plot saved as: {output_filename}")

    if pns_ok:
        print("PNS check passed successfully: maximum PNS level =", 100*pns_n.max(), "%")
        plt.show()
    else:
        print("PNS check failed: maximum PNS level =", 100*pns_n.max(), "%")
        plt.show()
        raise RuntimeError("PNS check failed!")

def check_acoustic_resonances(seq, gradFile, seq_file, scanner):
    """
    Plots the acoustic resonances of the sequence against the forbidden frequencies specified in gradFile
    """
    gradient_file = os.path.join(os.path.dirname(__file__), gradFile) # Path to the Prisma gradient file

    if os.path.exists(gradient_file):
        # Use gradient file if available
        from pypulseq.utils.siemens.readasc import readasc
        asc, extra = readasc(gradient_file) # Parse the ASC file

        # Extract acoustic resonances
        frequencies = asc["asGPAParameters"][0]["sGCParameters"]["aflAcousticResonanceFrequency"]
        bandwidths  = asc["asGPAParameters"][0]["sGCParameters"]["aflAcousticResonanceBandwidth"]
        #print("Frequencies:", frequencies)
        #print("Bandwidths:", bandwidths)

        # Combine frequencies and bandwidths into a list of dictionaries
        acoustic_resonances = [
            {"frequency": freq, "bandwidth": bw}
            for freq, bw in zip(frequencies.values(), bandwidths.values()) # Use .values() to get the dictionary values
            if freq != 0  # Exclude entries with zero frequency
        ]
        print(f"Using acoustic resonances from gradient file: {gradFile}")
    else:
        # Use default acoustic resonances if gradient file is not available
        if scanner == '3T':
            acoustic_resonances = DEFAULT_ACOUSTIC_RESONANCES_3T
        elif scanner == '7T':
            acoustic_resonances = DEFAULT_ACOUSTIC_RESONANCES_7T
        else:
            raise ValueError("Unknown scanner type. Please specify '3T' or '7T'.")
        print(f"Warning: Gradient file '{gradFile}' not found. Using default acoustic resonances.")

    # Extract MD5 hash for the title
    md5_hash = extract_md5_hash(seq_file)

    seq.calculate_gradient_spectrum(acoustic_resonances=acoustic_resonances, use_derivative=False, frequency_oversampling=10, window_width=(min(0.02,seq.duration()[0])))

    # Add title with MD5 hash
    plt.suptitle(f"Acoustic Resonances - Hash: {md5_hash}", fontsize=14, fontweight='bold')

    # Save the plot as PNG
    output_filename = f"acoustic_resonances_{os.path.basename(seq_file).replace('.seq', '')}.png"
    plt.savefig(output_filename, dpi=300, bbox_inches='tight')
    print(f"Acoustic resonances plot saved as: {output_filename}")
    plt.show()

def safety_check(seq_file, scanner='3T'):
    """
    Test a pulseq sequence for PNS and Acoustic Resonances
    """
    gradFile = "MP_GPA_K2309_2250V_951A_AS82.asc" # Siemens 3T Prisma gradient file for PNS and acoustic resonance check (place in the same folder as this file)

    # Create a Sequence object
    seq = pp.Sequence()

    # Print the sequnece duration
    #print(f"Sequence duration: {seq.duration()[0]:.3f} s")

    # Load the sequence from a .seq file
    seq.read(seq_file)

    # Check the PNS level
    check_PNS(seq, gradFile, seq_file)

    # Check the acoustic resonances
    check_acoustic_resonances(seq, gradFile, seq_file, scanner)

if __name__ == "__main__":
    if len(sys.argv) != 2 and len(sys.argv) != 3:
        print("Usage: python safety_check.py <path_to_seq_file> <scanner_type (3T or 7T, default 3T)>")
        sys.exit(1)

    seq_file = sys.argv[1]
    safety_check(seq_file)

File: safety_check.m
% This file will run a fast pns and forbidden resonances check and print a
% png file. If the asc file from the scanner is not found, the pns check is
% skipped.  

% USAGE:
% Include the pulseq-master path with subfolders in your matlab path
% Run the safety check as: safety_check('my_sequence.seq')
% If you have the asc file, place it in the same folder as this file and
% change the name below to match that of your file.


function safety_check(seq_file,scanner)

    % Gradient system file (must be in the same folder)
    gradFile = 'MP_GPA_K2309_2250V_951A_AS82.asc';
    
    f = figure('Color','w');%("Position",[680   558   990   420]); %Adjust accordingly if needed
    t = tiledlayout(1,2);

    % Load sequence
    seq = mr.Sequence();
    seq.read(seq_file);

    % PNS check
    check_PNS(seq, gradFile,t);

    % Acoustic resonance check
    check_acoustic_resonances(seq, gradFile, t, scanner);

    % Add a global title with the hash (if found)
    title(t, sprintf('Hash: %s', seq.signatureValue), ...
        'FontWeight','bold','Interpreter','none');


    [~,name] = fileparts(seq_file);
    saveas(f,[name '_safety_check.png']);

    disp('Safety check finished')
end

function check_PNS(seq, gradFile,t)
    nexttile(t);
    
    if exist(gradFile, 'file')
        [pns_ok, pns_n, pns_comp, ~] = seq.calcPNS(gradFile,false);

        safe_plot(pns_comp'*100, seq.gradRasterTime);
    
        if pns_ok
            fprintf('PNS check passed: max PNS level = %.1f %%\n', 100*max(pns_n));
        else
            fprintf('PNS check failed: max PNS level = %.1f %%\n', 100*max(pns_n));
            error('PNS check failed!');
        end
    else
        warning('Gradient file not found. Skipping PNS check.');
        
        % Create an empty plot to keep layout consistent
        plot(0,0,'w'); axis off;
        title('PNS Check (Skipped)');
    end
end

function check_acoustic_resonances(seq, gradFile,t,scanner)

    if strcmp(scanner,'3T')
        DEFAULT_ACOUSTIC_RESONANCES = [ ...
            struct('frequency', 590,  'bandwidth', 100); ...
            struct('frequency', 1140, 'bandwidth', 220) ...
        ];
    elseif strcmp(scanner,'7T')
        DEFAULT_ACOUSTIC_RESONANCES = [ ...
            struct('frequency', 1100,  'bandwidth', 300); ...
            struct('frequency', 550, 'bandwidth', 100) ...
        ];
    end

    if exist(gradFile, 'file')
        asc = mr.Siemens.readasc(gradFile);
        freqs = asc.asGPAParameters.sGCParameters.aflAcousticResonanceFrequency;
        bws   = asc.asGPAParameters.sGCParameters.aflAcousticResonanceBandwidth;
        
        acoustic_resonances = [];
        for k = 1:numel(freqs)
            if freqs(k) ~= 0
                acoustic_resonances(end+1).frequency = freqs(k);
                acoustic_resonances(end).bandwidth = bws(k);
            end
        end
    else
        warning('Gradient file not found. Using default acoustic resonances.');
        acoustic_resonances = DEFAULT_ACOUSTIC_RESONANCES;
    end

    [spectrograms, spectrogram_rss, frequencies] = calculate_gradient_spectrum(seq, acoustic_resonances, false, 10, 0.02, 2000, 'max', [], false);
    
    nexttile(t);
    hold all;
    plot(frequencies, spectrograms{1},'DisplayName','x');
    plot(frequencies, spectrograms{2},'DisplayName','y');
    plot(frequencies, spectrograms{3},'DisplayName','z');
    plot(frequencies, spectrogram_rss,'r','LineWidth',1,'DisplayName','rss');

    % Add acoustic resonances
    for k=1:numel(acoustic_resonances)
        res = acoustic_resonances(k);
        xline(res.frequency,'k-','LineWidth',2,'HandleVisibility','off');
        xline(res.frequency-res.bandwidth/2,'k--','LineWidth',2,'HandleVisibility','off');
        xline(res.frequency+res.bandwidth/2,'k--','LineWidth',2,'HandleVisibility','off');
    end
    xlabel('Frequency (Hz)'); ylabel('Amplitude');
    box on;
    legend();
    title('Acoustic Resonances');
end

